exporters/pelorus/deserialization/__init__.py

```diff
@@ -0,0 +1,392 @@
+# Copyright Red Hat
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+"""
+Declarative structuring and typechecking.
+
+Useful for checking that all necessary aspects of a structure from OpenShift are present and correct.
+
+Supported types:
+- some primitives (int, float, str, bool)
+- attrs classes (with nested supported types)
+- dicts (with nested supported value types)
+- lists (with nested supported types)
+- optionals of the above
+"""
+from __future__ import annotations
+
+import typing
+from typing import (
+    Any,
+    Iterable,
+    Mapping,
+    MutableMapping,
+    MutableSequence,
+    Optional,
+    TypeVar,
+    cast,
+    overload,
+)
+
+import attr
+import attrs
+
+from pelorus.utils import BadAttributePathError, get_nested
+from pelorus.utils._attrs_compat import NOTHING
+
+from .errors import (
+    DeserializationErrors,
+    FieldError,
+    FieldTypeCheckError,
+    InnerFieldDeserializationErrors,
+    MissingFieldError,
+    TypeCheckError,
+)
+
+# metadata
+_NESTED_PATH_KEY = "__pelorus_structure_nested_path"
+
+T = TypeVar("T")
+
+PRIMITIVE_TYPES: tuple[type, ...] = (int, float, str, bool)
+
+
+def nested(path: str) -> dict[str, Any]:
+    "Put the nested data path in a metadata dict."
+    return {_NESTED_PATH_KEY: path}
+
+
+def _is_attrs_class(cls: type) -> Optional[type["attr.AttrsInstance"]]:
+    """
+    Return the class if it is an attrs class, else None.
+
+    This exists because `attrs.has` is not a proper TypeGuard.
+    """
+    return cast("type[attr.AttrsInstance]", cls) if attrs.has(cls) else None
```

Comment:
Where the `attr.AttrsInstance" comes from? It's not part of the attr (at least in my venv). Does attr need minimum version which includes AttrsInstance ?